File: ts/src/discord.ts
export const DISCORD_API_BASE = 'https://discord.com/api/v10';

const MAX_RETRIES = 3;

let sleepFn = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

function sleep(ms: number): Promise<void> {
  return sleepFn(ms);
}

function getMessagesUrl(channelId: string): string {
  return `${DISCORD_API_BASE}/channels/${channelId}/messages`;
}

interface DiscordMessagePayload {
  content: string;
  embeds?: unknown[];
}

export async function sendDiscordMessage(
  token: string,
  channelId: string,
  text: string,
  fetchFn: typeof fetch = globalThis.fetch,
): Promise<Record<string, unknown>> {
  const url = getMessagesUrl(channelId);
  const payload: DiscordMessagePayload = { content: text };
  return postMessage(token, url, payload, fetchFn);
}

export async function sendDiscordCard(
  token: string,
  channelId: string,
  text: string,
  card: unknown,
  fetchFn: typeof fetch = globalThis.fetch,
): Promise<Record<string, unknown>> {
  const url = getMessagesUrl(channelId);
  const embeds = Array.isArray(card) ? card : [card];
  const payload: DiscordMessagePayload = { content: text, embeds };
  return postMessage(token, url, payload, fetchFn);
}

async function postMessage(
  token: string,
  url: string,
  payload: DiscordMessagePayload,
  fetchFn: typeof fetch,
): Promise<Record<string, unknown>> {
  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    const resp = await fetchFn(url, {
      method: 'POST',
      headers: {
        Authorization: `Bot ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (resp.status === 429) {
      if (attempt < MAX_RETRIES) {
        const retryAfter = resp.headers.get('Retry-After') || '1';
        const delaySecs = parseInt(retryAfter, 10) || 1;
        await sleep(delaySecs * 1000);
        continue;
      }
    }

    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`HTTP ${resp.status}: ${body}`);
    }
    return (await resp.json()) as Record<string, unknown>;
  }

  // This should be unreachable, but TypeScript needs it
  throw new Error('Unexpected state in postMessage retry loop');
}

/** Override sleep function for testing */
export function _setSleepFn(fn: (ms: number) => Promise<void>): void {
  sleepFn = fn;
}

File: go/discord.go
package wgrok

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"
)

const (
	DiscordAPIBase = "https://discord.com/api/v10"
)

// Package-level URL, overridable in tests.
var (
	DiscordChannelMessagesURL = func(channelID string) string {
		return DiscordAPIBase + "/channels/" + channelID + "/messages"
	}
)

type discordMessagePayload struct {
	Content string        `json:"content"`
	Embeds  []interface{} `json:"embeds,omitempty"`
}

// SendDiscordMessage sends a text-only Discord message to a channel.
func SendDiscordMessage(token, channelID, text string, client *http.Client) (map[string]interface{}, error) {
	payload := discordMessagePayload{Content: text}
	return postDiscordMessage(token, channelID, payload, client)
}

// SendDiscordCard sends a Discord message with an embed.
func SendDiscordCard(token, channelID, text string, card interface{}, client *http.Client) (map[string]interface{}, error) {
	var embeds []interface{}
	if embedList, ok := card.([]interface{}); ok {
		embeds = embedList
	} else {
		embeds = []interface{}{card}
	}
	payload := discordMessagePayload{Content: text, Embeds: embeds}
	return postDiscordMessage(token, channelID, payload, client)
}

func postDiscordMessage(token, channelID string, payload discordMessagePayload, client *http.Client) (map[string]interface{}, error) {
	if client == nil {
		client = http.DefaultClient
	}
	url := DiscordChannelMessagesURL(channelID)
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("marshal message payload: %w", err)
	}

	for attempt := 0; attempt <= MaxRetries; attempt++ {
		req, err := http.NewRequest("POST", url, bytes.NewReader(body))
		if err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "Bot "+token)
		req.Header.Set("Content-Type", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("send message: %w", err)
		}

		// Check for 429 Too Many Requests
		if resp.StatusCode == http.StatusTooManyRequests {
			io.Copy(io.Discard, resp.Body)
			resp.Body.Close()
			if attempt < MaxRetries {
				retryAfter := 1
				if retryAfterStr := resp.Header.Get("Retry-After"); retryAfterStr != "" {
					if parsed, err := strconv.Atoi(retryAfterStr); err == nil {
						retryAfter = parsed
						if retryAfter > 300 {
							retryAfter = 300
						}
					}
				}
				time.Sleep(time.Duration(retryAfter) * time.Second)
				continue
			}
			return nil, fmt.Errorf("HTTP %d: rate limited after %d retries", resp.StatusCode, MaxRetries)
		}

		defer resp.Body.Close()
		return readJSONResponse(resp)
	}
	return nil, fmt.Errorf("unreachable")
}
